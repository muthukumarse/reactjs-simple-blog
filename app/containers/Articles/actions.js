/*
 * Articles Actions
 *
*/

import {
  LOAD_ARTICLES,
  LOAD_ARTICLES_SUCCESS,
  LOAD_ARTICLES_ERROR,
} from './constants';

export function loadArticles() {
  return {
    type: LOAD_ARTICLES,
  };
}

export function articlesLoaded(articles) {
  return {
    type: LOAD_ARTICLES_SUCCESS,
    payload: { articles },
  };
}

export function articlesLoadingError(error) {
  return {
    type: LOAD_ARTICLES_ERROR,
    payload: { error },
  };
}
