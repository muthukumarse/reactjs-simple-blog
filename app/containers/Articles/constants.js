/*
 * Articles Constants
 */

export const LOAD_ARTICLES = 'boilerplate/Articles/LOAD_ARTICLES';
export const LOAD_ARTICLES_SUCCESS =
  'boilerplate/Articles/LOAD_ARTICLES_SUCCESS';
export const LOAD_ARTICLES_ERROR = 'boilerplate/Articles/LOAD_ARTICLES_ERROR';
