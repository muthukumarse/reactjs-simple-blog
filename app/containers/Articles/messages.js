/*
 * Articles Messages
 */
import { defineMessages } from 'react-intl';

export const scope = 'app.containers.Articles';

export default defineMessages({
  header: {
    id: `${scope}.header`,
    defaultMessage: 'All Articles',
  },
  articles: {
    id: `${scope}.articles`,
    defaultMessage: '{totalArticles} Articles',
  },
  pageSize: {
    id: `${scope}.pageSize`,
    defaultMessage: '{pageSize} per page',
  },
  previous: {
    id: `${scope}.previous`,
    defaultMessage: 'Previous',
  },
  next: {
    id: `${scope}.next`,
    defaultMessage: 'Next',
  },
});
