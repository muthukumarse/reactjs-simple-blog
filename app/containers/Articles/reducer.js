/*
 * Articles Reducer
 */
import { fromJS } from 'immutable';

import {
  LOAD_ARTICLES,
  LOAD_ARTICLES_SUCCESS,
  LOAD_ARTICLES_ERROR,
} from './constants';

// The initial state of the App
export const initialState = fromJS({
  articles: {},
  totalArticles: 0,
  error: false,
  errorDesc: '',
});

function articlesReducer(state = initialState, action) {
  switch (action.type) {
    case LOAD_ARTICLES:
      return initialState;
    case LOAD_ARTICLES_SUCCESS: {
      const {
        payload: { articles },
      } = action;
      const totalArticles = articles.length;
      return state.merge({ articles, totalArticles, error: false });
    }
    case LOAD_ARTICLES_ERROR: {
      const {
        payload: { error },
      } = action;
      return state.merge({
        articles: {},
        totalArticles: 0,
        error: true,
        errorDesc: error,
      });
    }
    default:
      return state;
  }
}

export default articlesReducer;
