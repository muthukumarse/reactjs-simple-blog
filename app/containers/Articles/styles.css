.articles-section {
  bottom: 0px;
  margin: 0 auto;
  padding: 10px 20px;
  padding-top: 0;
  overflow: hidden;
}

.article-head-block {
  position: absolute;
  top: 20px;
  right: 20px;
  width: 50%;
}

.article-head-block-section {
  width: 50%;
  display: inline-block;
}
.article-head-block-section-right {
  text-align: right;
}
.post-list > ul {
  display: grid;
  grid-template-columns: repeat(3, 1fr);
  grid-gap: 40px;
}
.post-list > ul {
  display: grid;
  grid-template-columns: repeat(2, 1fr);
  grid-gap: 20px;
  grid-auto-rows: 1fr;
}
.post-list ul {
  list-style: none;
  margin: 0;
  padding: 0;
}

.post-list {
  margin: 0 auto;
  max-width: 1000px;
}

.post {
  margin-bottom: 0;
}

.post {
  margin-bottom: 20px;
}

.post .post-link {
  background: #fff;
  box-shadow: 0 1px 1px 0 rgba(0, 0, 0, 0.05);
  border-radius: 4px;
  cursor: pointer;
  display: block;
  height: 100%;
  padding: 30px;
  position: relative;
}

a {
  color: #017eb6;
  text-decoration: none;
}

.post .post-title {
  color: #101010;
  display: block;
  font-size: 16px;
  font-weight: 600;
  margin-bottom: 5px;
}

.post .post-meta {
  color: #8b99a1;
  display: block;
  font-size: 13px;
  margin-bottom: 12px;
}

.post .post-link:after {
  content: '';
  background-color: #38c1fe;
  border-radius: 4px;
  display: block;
  opacity: 0;
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  transition: opacity 0.2s ease-in-out;
  z-index: 1;
}

.post .post-link:before {
  content: 'Edit Post';
  color: #fff;
  display: block;
  font-size: 18px;
  font-weight: 600;
  letter-spacing: 2px;
  opacity: 0;
  position: absolute;
  top: 50%;
  left: 50%;
  -webkit-transform: translate(-50%, -50%);
  transform: translate(-50%, -50%);
  text-transform: uppercase;
  transition: opacity 0.2s ease-in-out;
  z-index: 2;
}

.pagination-section {
  text-align: right;
  background: #f1f3f4;
}
.pagination {
  display: inline-block;
  padding-left: 0;
  margin: 20px 0;
  border-radius: 4px;
}

.pagination > li {
  display: inline;
}

.pagination > li:first-child > a,
.pagination > li:first-child > span {
  margin-left: 0;
  border-bottom-left-radius: 4px;
  border-top-left-radius: 4px;
}

.pagination > li > a,
.pagination > li > span {
  position: relative;
  float: left;
  padding: 6px 12px;
  line-height: 1.42857143;
  text-decoration: none;
  color: #000000;
  margin-left: -1px;
}

.pagination > li > a:hover {
  background-color: #41addd;
}

.pagination > .active > a,
.pagination > .active > a:focus,
.pagination > .active > a:hover,
.pagination > .active > span,
.pagination > .active > span:focus,
.pagination > .active > span:hover {
  z-index: 3;
  color: #000000;
  background-color: #ffffff;
  border-color: #337ab7;
  border-bottom: 1px solid #000;
  cursor: default;
}

.post .post-link {
  background: #fff;
  box-shadow: 0 1px 1px 0 rgba(0, 0, 0, 0.05);
  border-radius: 4px;
  cursor: pointer;
  display: block;
  height: 100%;
  padding: 30px;
  position: relative;
}
.post .post-link:before {
  content: 'View Post';
  color: #fff;
  display: block;
  font-size: 18px;
  font-weight: 600;
  letter-spacing: 2px;
  opacity: 0;
  position: absolute;
  top: 50%;
  left: 50%;
  -webkit-transform: translate(-50%, -50%);
  transform: translate(-50%, -50%);
  text-transform: uppercase;
  transition: opacity 0.2s ease-in-out;
  z-index: 2;
}
.post .post-link:hover:before {
  opacity: 1;
}
.post .post-link:after {
  content: '';
  background-color: #38c1fe;
  border-radius: 4px;
  display: block;
  opacity: 0;
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  transition: opacity 0.2s ease-in-out;
  z-index: 1;
}
.post .post-link:hover:after {
  opacity: 0.95;
}
