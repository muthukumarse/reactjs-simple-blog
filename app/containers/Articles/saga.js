/**
 * Articles Saga to get Data from Github
 */

import { call, put, takeLatest } from 'redux-saga/effects';
import request from 'utils/request';
import { articlesLoaded, articlesLoadingError } from './actions';
import { LOAD_ARTICLES } from './constants';

export function* getArticles() {
  const requestURL = `https://jsonplaceholder.typicode.com/posts`;

  try {
    const articles = yield call(request, requestURL);
    yield put(articlesLoaded(articles));
  } catch (err) {
    yield put(articlesLoadingError(err));
  }
}

export default function* getArticlesList() {
  yield takeLatest(LOAD_ARTICLES, getArticles);
}
