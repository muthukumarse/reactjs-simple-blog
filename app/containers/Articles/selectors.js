/**
 * Articles selectors
 */

import { createSelector } from 'reselect';
import { initialState } from './reducer';

const selectArticles = state => state.get('articles', initialState);

const makeGetArticles = () =>
  createSelector(selectArticles, articlesState => articlesState.toJS());

export default makeGetArticles;
export { selectArticles, makeGetArticles };
