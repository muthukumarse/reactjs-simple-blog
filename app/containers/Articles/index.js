/*
 * Articles
 */

import React from 'react';
import { injectIntl, intlShape, FormattedMessage } from 'react-intl';
import { connect } from 'react-redux';
import { compose } from 'redux';
import { createStructuredSelector } from 'reselect';
import injectReducer from 'utils/injectReducer';
import injectSaga from 'utils/injectSaga';
import Pagination from 'react-js-pagination';

import messages from './messages';
import { makeGetArticles } from './selectors';
import { loadArticles } from './actions';
import reducer from './reducer';
import saga from './saga';
import './styles.css';

export class Articles extends React.Component {
  constructor(props, context) {
    super(props, context);
    this.state = {
      activePage: 1,
      pageSize: 10,
      articleStartIndex: 0,
      articleEndIndex: 9,
    };
  }

  componentDidMount() {
    this.props.loadArticles();
  }

  onPageChange = activePage => {
    const { pageSize } = this.state;
    const articleStartIndex = pageSize * (activePage - 1);
    const articleEndIndex = pageSize * activePage - 1;
    this.setState({
      activePage,
      articleStartIndex,
      articleEndIndex,
    });
    // console.log('activePage', activePage);
  };

  onPageSizeChange = evnt => {
    const { activePage } = this.state;
    const pageSize = evnt.target.value;
    const articleStartIndex = 0;
    const articleEndIndex = pageSize * activePage - 1;
    this.setState({
      pageSize,
      activePage: 1,
      articleStartIndex,
      articleEndIndex,
    });
    // console.log('pageSize', evnt.target.value);
  };

  render() {
    // console.log(this.props.articlesState.articles.length)
    const {
      articlesState: { articles, totalArticles },
      intl: { formatMessage },
    } = this.props;
    const {
      activePage,
      pageSize,
      articleStartIndex,
      articleEndIndex,
    } = this.state;

    if (totalArticles === 0) return <div id="loading">Loading ...</div>;

    const articleList = articles.slice(articleStartIndex, articleEndIndex + 1);
    // console.log(articleStartIndex, articleEndIndex);

    return (
      <div style={{ padding: '1em' }}>
        <div className="articles-section">
          <div className="article-head-block">
            <div className="article-head-block-section">
              <FormattedMessage {...messages.header} /> ({totalArticles})
            </div>
            <div className="article-head-block-section article-head-block-section-right">
              <select id="pageSize" onChange={this.onPageSizeChange}>
                <option value="10">
                  {formatMessage(messages.pageSize, { pageSize: 10 })}
                </option>
                <option value="25">
                  {formatMessage(messages.pageSize, { pageSize: 25 })}
                </option>
                <option value="50">
                  {formatMessage(messages.pageSize, { pageSize: 50 })}
                </option>
                <option value="100">
                  {formatMessage(messages.pageSize, { pageSize: 100 })}
                </option>
              </select>
            </div>
          </div>
          <div className="post-list">
            <ul>
              {articleList.map(article => (
                <li
                  key={`article-${article.id}`}
                  id={`post-${article.id}`}
                  className="post"
                >
                  <a
                    href={`/articles/post/${article.id}`}
                    className="post-link"
                  >
                    <span className="h2 post-title">{article.title}</span>
                    <span className="post-meta">
                      <p>{article.body.substr(0, 60)} ...</p>
                    </span>
                  </a>
                </li>
              ))}
            </ul>
          </div>
          <div className="pagination-section">
            <Pagination
              hideFirstLastPages
              pageRangeDisplayed={10}
              activePage={activePage}
              itemsCountPerPage={pageSize}
              totalItemsCount={totalArticles}
              onChange={this.onPageChange}
              prevPageText="Previous Page"
              nextPageText="Next Page"
            />
          </div>
        </div>
      </div>
    );
  }
}

Articles.propTypes = {
  intl: intlShape,
};

export function mapDispatchToProps(dispatch) {
  return {
    loadArticles: () => dispatch(loadArticles()),
  };
}

const mapStateToProps = createStructuredSelector({
  articlesState: makeGetArticles(),
});

const withConnect = connect(
  mapStateToProps,
  mapDispatchToProps,
);

const withReducer = injectReducer({ key: 'articles', reducer });
const withSaga = injectSaga({ key: 'articles', saga });

export default compose(
  withReducer,
  withSaga,
  withConnect,
)(injectIntl(Articles));
