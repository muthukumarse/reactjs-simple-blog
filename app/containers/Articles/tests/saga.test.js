/**
 * Tests for Articles sagas
 */

import { put, takeLatest } from 'redux-saga/effects';

import { LOAD_ARTICLES } from 'containers/Articles/constants';
import { articlesLoadingError } from 'containers/Articles/actions';

import getArticlesList, { getArticles } from '../saga';

describe('getArticles Saga', () => {
  let getArticlesGenerator;

  beforeEach(() => {
    getArticlesGenerator = getArticles();

    const selectDescriptor = getArticlesGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot();

    const callDescriptor = getArticlesGenerator.next().value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should call the error action if the response errors', () => {
    const response = new Error('Some error');
    const putDescriptor = getArticlesGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(articlesLoadingError(response)));
  });
});

describe('getArticlesList Saga', () => {
  const getArticlesListSaga = getArticlesList();

  it('should start task to watch for LOAD_ARTICLES action', () => {
    const takeLatestDescriptor = getArticlesListSaga.next().value;
    expect(takeLatestDescriptor).toEqual(
      takeLatest(LOAD_ARTICLES, getArticles),
    );
  });
});
