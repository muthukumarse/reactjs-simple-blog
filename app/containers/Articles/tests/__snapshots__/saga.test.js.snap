// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getArticles Saga should call the error action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "https://jsonplaceholder.typicode.com/posts",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`getArticles Saga should call the error action if the response errors 2`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "payload": Object {
        "articles": undefined,
      },
      "type": "boilerplate/Articles/LOAD_ARTICLES_SUCCESS",
    },
    "channel": null,
  },
}
`;
